### ISSUE TITLE
OpenSSL should be replaced by GnuTLS

### ISSUE THREAD

**Issue by rkday** [SIPp/sipp#32]
OpenSSL is not GPL-licensed, which causes problems for SIPp distributors. We should use GnuTLS (http://www.gnutls.org/) for handling TLS mode instead.

**Comment by linuxmaniac** [comment]
After 4 years, instead of porting to GnuTLS can you please add the license exception?
```
 * In addition, as a special exception, the copyright holders give
 * permission to link the code of portions of this program with the
 * OpenSSL library under certain conditions as described in each
 * individual source file, and distribute linked combinations
 * including the two.
 * You must obey the GNU General Public License in all respects
 * for all of the code used other than OpenSSL.  If you modify
 * file(s) with this exception, you may extend this exception to your
 * version of the file(s), but you are not obligated to do so.  If you
 * do not wish to do so, delete this exception statement from your
 * version.  If you delete this exception statement from all source
 * files in the program, then also delete it here.
```
https://people.gnome.org/~markmc/openssl-and-the-gpl.html

**Comment by wdoekes** [comment]
Where do we add this?

**Comment by wdoekes** [comment]
Something like that?

**Comment by rkday** [comment]
@wdoekes , pretty sure you can't add that license exception - you're not the copyright holder for most of the code, so can't allow the code to be used in ways the original authors haven't granted permission for.

(In theory it would be possible to conact all the authors and get their permission, but that's a fairly substantial undertaking.)

**Comment by wdoekes** [comment]
Hi Rob, thanks for pitching in.

You're probably right on that.

So. What are our options, apart from migrating to GnuTLS? Especially when we consider that we would be hard-pressed to find someone against this exception, right?

Would it be sufficient to contact the SIPp mailing list and request any original author to speak up if they are against? That, combined with an attempt to hunt down the most prolific authors of SIPp code could be enough to convince those with rights to agree that a decent attempt has been undertaken to respect everyones rights.

**Comment by uhle** [comment]
Instead of porting the code to use GnuTLS, [WolfSSL](/wolfSSL/wolfssl) could be used with only few changes. Its OpenSSL portability layer is far more advanced and its license is GPL2 compliant. If you're interested, you could check out and try branch [wolfssl-support](/uhle/sipp/tree/wolfssl-support) or have a look at sipp@851d1fdbca795ebea5c646512695a723789f7d19.

I could also make a pull request if anyone wants me to.

**Comment by wdoekes** [comment]
Yes please!

I'll make sure it gets merged ASAP.

**Comment by wdoekes** [comment]
Fixed with PR #505.

**Comment by lemenkov** [comment]
Awesome!
